src/beamlime/constructors/injection.py

```diff
@@ -0,0 +1,171 @@
+        self._providers[_product_type] = new_provider_call
+
+    def _validate_new_provider(
+        self, product_type: ProductType, new_provider: Provider
+    ) -> None:
+        """
+        Raises
+        ------
+        ProviderExistsError:
+            If there is an existing provider of the ``product_type``.
+            However, if the full name of the ``new_provider_call``
+            is same as the existing one, it will register
+            the new callable object as the provider.
+
+        MismatchingProductTypeError:
+            If the return type of the ``provider_call`` is
+            incompatible with ``product_type``.
+            The return type of the ``provider_call`` should be a subclass of
+            ``product_type``or the same object as ``product_type``.
+            See ``ProviderCall.is_supprted_type``.
+        """
+        if product_type in self._providers:
+            existing_provider = self._providers[product_type]
+            if existing_provider != new_provider:
+                raise ProviderExistsError(
+                    f"Can not register ``{new_provider.call_name}``."
+                    f"Provider for ``{product_type.__name__}`` "
+                    f"already exist. ``{existing_provider.call_name}``"
+                )
+        if not new_provider.can_provide(product_type):
+            raise MismatchingProductTypeError(
+                f"{product_type} can not be provided by " f"{new_provider.constructor}"
+            )
+
+    def provider(self, callable_obj: Constructor):
+        """
+        Register the decorated provider function into this factory.
+        """
+        from .inspectors import UnknownType
+
+        self.register(UnknownType, callable_obj)
+        return callable_obj
+
+
+class Empty:
+    ...
+
+
+class InjectionInterface(RegistrationInterface):
+    """Dependency injection interfaces."""
+
+    def build_arg_dependency(self, arg_spec: DependencySpec):
+        try:
+            return self.assemble(arg_spec.product_type)
+        except ProviderNotFoundError as err:
+            from inspect import Signature
+
+            from .inspectors import UnknownType
+
+            if (arg_spec.default_product != Signature.empty) or (
+                arg_spec.product_type == UnknownType
+            ):
+                return Empty
+            raise err
+
+    def build_arguments(self, provider: Provider):
+        return {
+            arg_name: _product
+            for arg_name, arg_spec in provider.arg_dep_specs.items()
+            if (_product := self.build_arg_dependency(arg_spec)) != Empty
+        }
+
+    def build_attr_dependency(self, dep_type: ProductType):
+        try:
+            return self.assemble(dep_type)
+        except ProviderNotFoundError:
+            return Empty
+
+    def inject_dependencies(self, _obj: _Product) -> _Product:
+        """
+        Check if the ``incomplete_obj`` has attribute dependencies to be filled.
```

Comment:
What is `incomplete_obj`? Did you rename the arg and forget to update the comment?